### PR TITLE
[SWDEV-274378][LRN] Fix in local response normalization layer to synchronize local memory access within workgroup

### PR DESCRIPTION
**Pull request by muralinr**
Fixed a NAN issue by providing a fix to synchronize local memory access within workgroup in LRN (local response normalization) layer.

### CODE REVIEW

**Comment by muralinr** [comment]
Fixed a NAN issue for this failed config.
root@ixt-rack-185:/testx/MIOpen/build# ./bin/MIOpenDriver lrn -H 7 -W 7 -n 1 -c 1056 -N 4 -K 1 -A 1 -B 1 -m within -F 0 -t 1
MIOpenDriver lrn -H 7 -W 7 -n 1 -c 1056 -N 4 -K 1 -A 1 -B 1 -m within -F 0 -t 1
GPU Kernel Time Forward LRN Elapsed: 0.038920 ms
Forward LRN Verifies on CPU and GPU (err=0.000000)
GPU Kernel Time Backward LRN Elapsed: 0.036999 ms
Backward LRN Verifies on CPU and GPU (err=0.000000)

File: src/kernels/MIOpenLRNBwd.cl [hunk 1]
```diff
@@ -161,6 +161,7 @@ MIOpenLRNWithinChannelBwd(const __global _FLOAT* top,
         }
     }
 
+    barrier(CLK_LOCAL_MEM_FENCE);
```
Comment (junliume) [inline src/kernels/MIOpenLRNBwd.cl:164]: @muralinr May I ask why it fixes the issue? Looks like a racing condition but where? :-(

P.S. a comment for Artem: Please add the config to the unit test too. Thanks! ;-P

**Review by atamazov — CHANGES_REQUESTED** [review]
Regression test to be added.

File: src/kernels/MIOpenLRNBwd.cl (hunk 1 above)
Comment (atamazov) [inline src/kernels/MIOpenLRNBwd.cl:164]: - Do we need this fix for wavesize32 devices (Navi21) only?
- If we apply it to _all_ targets, does it affect performance?

**Comment by junliume** [comment]
@muralinr could you please update the ctest? Hope we can include this one in the next staging today. Thanks!

**Comment by muralinr** [comment]
> @muralinr could you please update the ctest? Hope we can include this one in the next staging today. Thanks!

Hi Jun, LRN ctest is different from it's driver test.  I am not able to reproduce it currently on ctest for both FP32 and FP16.  I will try to reproduce it and add it once I reproduce it.

root@ixt-rack-185:/testx/MIOpen/build# ./bin/test_lrn_test --half -H 7 -W 7 -n 1 -c 1056 -N 4 -A 1 -B 1 -K 1 -F 0 --mode Within_Channel
root@ixt-rack-185:/testx/MIOpen/build# ./bin/test_lrn_test --float -H 7 -W 7 -n 1 -c 1056 -N 4 -A 1 -B 1 -K 1 -F 0 --mode Within_Channel
root@ixt-rack-185:/testx/MIOpen/build#

**Comment by atamazov** [comment]
@muralinr 
> LRN ctest is different from it's driver test. I am not able to reproduce it currently on ctest for both FP32 and FP16.

Yes, I tried this couple of days ago and found that reproducing the failure with test_lrn_test is not trivial at least (if even possible). This may require modification of test_lrn_test.

**Comment by atamazov** [comment]
Maybe we need help from @ce1adon

**Comment by ce1adon** [comment]
> Maybe we need help from @ce1adon

Talked to @muralinr . Could be inconsistency between driver and ctest. Murali is working on aligning the arguments and flags.

**Comment by atamazov** [comment]
@ce1adon Great, thank you very much!

**Comment by junliume** [comment]
@muralinr  could you update this PR with a regression test? Hopefully this one can catch the next staging.

**Comment by muralinr** [comment]
> @muralinr could you update this PR with a regression test? Hopefully this one can catch the next staging.

Hi Jun, I am working on reproducing this issue on CTEST.  I am not yet successfully.  I will add this test as soon as I reproduce it.

**Comment by muralinr** [comment]
> ./bin/MIOpenDriver lrn -H 7 -W 7 -n 1 -c 1056 -N 4 -K 1 -A 1 -B 1 -m within -F 0 -t 1

Hi Jun,
I am able to reproduce this NAN issue with the following configs.

root@ixt-rack-185:/testx/MIOpen/build# ./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 0
verify_lrn_bwd
Input Tensor Y 1, 1056, 7, 7
Input Tensor DY 1, 1056, 7, 7
Input Tensor X 1, 1056, 7, 7
Max diff: 177.25
Mismatch at 1: 0.0777237 != 0.0777237
Non finite number found in gpu at 49709: -nan

**Comment by muralinr** [comment]
@junliume This test case is already part of 250 LRN tests ( ./bin/test_lrn_test --all). I am able to reproduce it when I run this command on Navi21.

root@ixt-rack-185:/testx/MIOpen/build# ./bin/test_lrn_test --all
./bin/test_lrn_test --float --input 1, 16, 4096, 4096 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: 0.000131807
Iteration: 62
verify_lrn_bwd
Input Tensor Y 1, 16, 4096, 4096
Input Tensor DY 1, 16, 4096, 4096
Input Tensor X 1, 16, 4096, 4096
Max diff: 114.75
Mismatch at 1: 0.0777237 != 0.0777237
./bin/test_lrn_test --float --input 1, 480, 128, 256 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: 0.000501433
Iteration: 194
verify_lrn_bwd
Input Tensor Y 1, 480, 128, 256
Input Tensor DY 1, 480, 128, 256
Input Tensor X 1, 480, 128, 256
Max diff: 131.052
Mismatch at 1: 0.0777237 != 0.0777237
./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 242
verify_lrn_bwd
Input Tensor Y 1, 1056, 7, 7
Input Tensor DY 1, 1056, 7, 7
Input Tensor X 1, 1056, 7, 7
Max diff: 177.25
Mismatch at 1: 0.0777237 != 0.0777237
Non finite number found in gpu at 49951: -nan
./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 5 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 244
verify_lrn_bwd
Input Tensor Y 1, 1056, 7, 7
Input Tensor DY 1, 1056, 7, 7
Input Tensor X 1, 1056, 7, 7
Max diff: 5.96046e-07
Mismatch at 1: -0.703801 != -0.703801
Non finite number found in gpu at 46226: -nan
./bin/test_lrn_test --float --input 1, 2048, 11, 11 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 248
verify_lrn_bwd
Input Tensor Y 1, 2048, 11, 11
Input Tensor DY 1, 2048, 11, 11
Input Tensor X 1, 2048, 11, 11
Max diff: 114.75
Mismatch at 1: 0.0777237 != 0.0777237
Non finite number found in gpu at 114501: -nan
./bin/test_lrn_test --float --input 1, 2048, 11, 11 --N 5 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 250
verify_lrn_bwd
Input Tensor Y 1, 2048, 11, 11
Input Tensor DY 1, 2048, 11, 11
Input Tensor X 1, 2048, 11, 11
Max diff: 112.477
Mismatch at 1: -0.703801 != -0.703801
Non finite number found in gpu at 175242: -nan

**Comment by junliume** [comment]
> @junliume This test case is already part of 250 LRN tests ( ./bin/test_lrn_test --all). I am able to reproduce it when I run this command on Navi21.
> 
> root@ixt-rack-185:/testx/MIOpen/build# ./bin/test_lrn_test --all ./bin/test_lrn_test --float --input 1, 16, 4096, 4096 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel FAILED: 0.000131807 Iteration: 62 verify_lrn_bwd Input Tensor Y 1, 16, 4096, 4096 Input Tensor DY 1, 16, 4096, 4096 Input Tensor X 1, 16, 4096, 4096 Max diff: 114.75 Mismatch at 1: 0.0777237 != 0.0777237 ./bin/test_lrn_test --float --input 1, 480, 128, 256 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel FAILED: 0.000501433 Iteration: 194 verify_lrn_bwd Input Tensor Y 1, 480, 128, 256 Input Tensor DY 1, 480, 128, 256 Input Tensor X 1, 480, 128, 256 Max diff: 131.052 Mismatch at 1: 0.0777237 != 0.0777237 ./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel FAILED: -nan Iteration: 242 verify_lrn_bwd Input Tensor Y 1, 1056, 7, 7 Input Tensor DY 1, 1056, 7, 7 Input Tensor X 1, 1056, 7, 7 Max diff: 177.25 Mismatch at 1: 0.0777237 != 0.0777237 Non finite number found in gpu at 49951: -nan ./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 5 --alpha 1 --beta 1 --K 1 --mode Within_Channel FAILED: -nan Iteration: 244 verify_lrn_bwd Input Tensor Y 1, 1056, 7, 7 Input Tensor DY 1, 1056, 7, 7 Input Tensor X 1, 1056, 7, 7 Max diff: 5.96046e-07 Mismatch at 1: -0.703801 != -0.703801 Non finite number found in gpu at 46226: -nan ./bin/test_lrn_test --float --input 1, 2048, 11, 11 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel FAILED: -nan Iteration: 248 verify_lrn_bwd Input Tensor Y 1, 2048, 11, 11 Input Tensor DY 1, 2048, 11, 11 Input Tensor X 1, 2048, 11, 11 Max diff: 114.75 Mismatch at 1: 0.0777237 != 0.0777237 Non finite number found in gpu at 114501: -nan ./bin/test_lrn_test --float --input 1, 2048, 11, 11 --N 5 --alpha 1 --beta 1 --K 1 --mode Within_Channel FAILED: -nan Iteration: 250 verify_lrn_bwd Input Tensor Y 1, 2048, 11, 11 Input Tensor DY 1, 2048, 11, 11 Input Tensor X 1, 2048, 11, 11 Max diff: 112.477 Mismatch at 1: -0.703801 != -0.703801 Non finite number found in gpu at 175242: -nan

Hope #1408 will be able to discover similar issues in the future.

**Review by junliume — APPROVED** [review]

**Comment by atamazov** [comment]
@junliume 
> Hope https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1408 will be able to discover similar issues in the future.

It won't because it updates the driver, not ctest. Ctest is already able to detect NaNs etc.

**Comment by atamazov** [comment]
@muralinr 
> This test case is already part of 250 LRN tests ( ./bin/test_lrn_test --all). I am able to reproduce it when I run this command on Navi21.

This can't be so, otherwise our CI would catch this error before this PR. Something is wrong.

**Comment by muralinr** [comment]
> @muralinr
> 
> > This test case is already part of 250 LRN tests ( ./bin/test_lrn_test --all). I am able to reproduce it when I run this command on Navi21.
> 
> This can't be so, otherwise our CI would have caught this bug before this PR. Something is wrong.

@atamazov I just ran LRN tests on develop.  NAN issue still shows up. We need to run all tests ( ./bin/test_lrn_test --all) instead of just one test (./bin/test_lrn_test).  Are you doing that?  Can you please log into Navi21 machine (ssh master@10.216.80.9) and try yourself? 
  
root@ixt-rack-185:/testx/test_develop/MIOpen/build# ./bin/test_lrn_test
root@ixt-rack-185:/testx/test_develop/MIOpen/build# ./bin/test_lrn_test --all
./bin/test_lrn_test --float --input 1, 16, 3072, 3072 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 56
verify_lrn_bwd
Input Tensor Y 1, 16, 3072, 3072
Input Tensor DY 1, 16, 3072, 3072
Input Tensor X 1, 16, 3072, 3072
Max diff: 114.75
Mismatch at 1: 0.0777237 != 0.0777237
Non finite number found in gpu at 83862723: -nan
./bin/test_lrn_test --float --input 1, 480, 64, 128 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: 0.00114812
Iteration: 188
verify_lrn_bwd
Input Tensor Y 1, 480, 64, 128
Input Tensor DY 1, 480, 64, 128
Input Tensor X 1, 480, 64, 128
Max diff: 177.25
Mismatch at 1: 0.0777237 != 0.0777237
./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 242
verify_lrn_bwd
Input Tensor Y 1, 1056, 7, 7
Input Tensor DY 1, 1056, 7, 7
Input Tensor X 1, 1056, 7, 7
Max diff: 40.2034
Mismatch at 1: 0.0777237 != 0.0777237
Non finite number found in gpu at 46713: -nan
./bin/test_lrn_test --float --input 1, 1056, 7, 7 --N 5 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 244
verify_lrn_bwd
Input Tensor Y 1, 1056, 7, 7
Input Tensor DY 1, 1056, 7, 7
Input Tensor X 1, 1056, 7, 7
Max diff: 58.5789
Mismatch at 1: -0.703801 != -0.703801
Non finite number found in gpu at 47593: -nan
./bin/test_lrn_test --float --input 1, 2048, 11, 11 --N 4 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 248
verify_lrn_bwd
Input Tensor Y 1, 2048, 11, 11
Input Tensor DY 1, 2048, 11, 11
Input Tensor X 1, 2048, 11, 11
Max diff: 114.75
Mismatch at 1: 0.0777237 != 0.0777237
Non finite number found in gpu at 113532: -nan
./bin/test_lrn_test --float --input 1, 2048, 11, 11 --N 5 --alpha 1 --beta 1 --K 1 --mode Within_Channel
FAILED: -nan
Iteration: 250
verify_lrn_bwd
Input Tensor Y 1, 2048, 11, 11
Input Tensor DY 1, 2048, 11, 11
Input Tensor X 1, 2048, 11, 11
Max diff: 171.02
Mismatch at 1: -0.703801 != -0.703801
Non finite number found in gpu at 163988: -nan
root@ixt-rack-185:/testx/test_develop/MIOpen/build#

**Comment by muralinr** [comment]
> @muralinr and @atamazov I just tested the issue with the newly released 5.0 and MIOpen develop branch, and I cannot reproduce this issue any more.
> 
> ```
> root@junliu-dev:~/MIOpen/build# ./bin/MIOpenDriver lrn -H 7 -W 7 -n 1 -c 1056 -N 4 -K 1 -A 1 -B 1 -m within -F 0 -t 1
> MIOpenDriver lrn -H 7 -W 7 -n 1 -c 1056 -N 4 -K 1 -A 1 -B 1 -m within -F 0 -t 1
> GPU Kernel Time Forward LRN Elapsed: 0.009880 ms
> Forward LRN Verifies on CPU and GPU (err=0.000000)
> GPU Kernel Time Backward LRN Elapsed: 0.012399 ms
> Backward LRN Verifies on CPU and GPU (err=0.000000)
> ```
> 
> The question is: do we still need to "fix" it? If the changes implemented here are indeed a fix I would vote for yes. If the changes are "preventative" and may have performance penalties, I would vote for no. What do you think?

@junliume This is a race condition.  It might show up in future.  I would still vote for fix.

**Comment by atamazov** [comment]
@muralinr Sounds reasonable

**Review by atamazov — APPROVED** [review]

**Comment by muralinr** [comment]
> @muralinr
> 
> > Do we need this fix for wavesize32 devices (Navi21) only?
> 
> The answer is NO, right?

@atamazov You are correct.

**Pull request merged**